%\VignetteIndexEntry{sequenza SNP-array usage example}
%\VignetteDepends{sequenza}
%\VignetteKeywords{LOH}
%\VignetteKeywords{Heterogenity}
%\VignetteKeywords{Cancer sequencing}
%\VignettePackage{sequenza}
%\VignettePackage{SNP}

\documentclass[10pt,a4paper]{article}
\usepackage{a4wide}
\usepackage{hyperref}
\usepackage{subfigure}
\usepackage{listings}
\usepackage{float}

\newcommand{\shellcmd}[1]{\\\indent\indent\texttt{\footnotesize\# #1}\\}
\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textit{#1}}}

\author{Francesco Favero\footnote{favero@cbs.dtu.dk}, Aron C. Eklund}

\begin{document}
\SweaveOpts{concordance=TRUE}
\title{sequenza possible SNP-array usage example}
\maketitle

\tableofcontents


\section{Working with SNP array data}
<<label=loadingSequenza, eval=TRUE>>=
library(sequenza)
@
\subsection{Preparing the data}
<<label=loadingFromCopynumber, eval=TRUE>>=
data(BAF)
data(logR)
@

<<label=SNPdataPrep, eval=TRUE>>=
sample.i <- data.frame(chromosome = BAF$chrs, n.base = BAF$pos, 
                      Bf = BAF$S1, adjusted.ratio = logR$S1,
                      depth.sample = 1, good.s.reads = 1,
                      ref.zygosity = 'hom', stringsAsFactors = FALSE)
@
\subsubsection{Correcting logR with a normal sample, or with the mean logR value}
Without a reference sample (normal germline sample) we can try to divide for the mean value. It would be correct to use the germline logR.
<<label=logRMeansub, eval=TRUE>>=
#sample.i$adjusted.ratio <- 2^(sample.i$adjusted.ratio)
#sample.i$adjusted.ratio <- sample.i$adjusted.ratio / mean(sample.i$adjusted.ratio)
sample.i$adjusted.ratio <- 2^(sample.i$adjusted.ratio/0.55)

@
\subsubsection{Retrieve the homozygous position}
It should be available a germline sample to get the heterozygours SNP, doing in the same sample it's a risk if the sample is pure. A threshold around 0.25 or 0.35 can be picked to subset the heterozygous position on the germline.
In the example we are lowering the threshold while taking the SNP from the same aberrant sample.
<<label=hetFind, eval=TRUE>>=
het.lim <- 0.2
is.het <- sample.i$Bf >= het.lim & sample.i$Bf <= 1 - het.lim
sample.i$ref.zygosity[is.het] <- 'het'
sample.i$Bf[sample.i$Bf >= 0.5] <- 1 - sample.i$Bf[sample.i$Bf >= 0.5]
sample.het.i <- sample.i[is.het, ]
@
\subsection{Windowing logR values.}
<<label=logRWin, eval=TRUE, results=hide>>=
 
snp.r.win <- windowValues(x = sample.i$adjusted.ratio,
                          positions = sample.i$n.base,
                          chromosomes = sample.i$chromosome,
                          window = 1e6, overlap = 1)
@
\subsection{Windowing B-allele frequencies values.}
<<label=BAFWin, eval=TRUE, results=hide>>=
snp.b.win <- windowValues(x = sample.het.i$Bf,
                          positions = sample.het.i$n.base,
                          chromosomes = sample.het.i$chromosome,
                          window = 1e6, overlap = 1)
@
\subsection{Chromosome view without mutation}
\begin{figure}[H]
      \centering
<<label=chrViewNoMut, fig=TRUE, width=6, height=6>>=

chromosome.view(baf.windows = snp.b.win[[1]], 
                ratio.windows = snp.r.win[[1]],
                 min.N.ratio = 1)
@ 
      \caption{Plots B-allele frequencies (top) and un-logged-logR (bottom) with SNP array data. }
      \label{fig:crhomViewSNP}
\end{figure}

\subsection{Segmenting with the \textit{copynumber} package}
<<label=copynumber, eval=TRUE>>=
breaks <- find.breaks(sample.het.i, gamma = 20, kmin = 15, baf.thres = c(0, 0.5))
seg.i  <- segment.breaks(sample.i, breaks = breaks)
@

\subsection{Using the Bayesian inference on segmented SNP arrays}
<<label=adjustSegs, eval=TRUE>>=

weights.snp   <- 150 + round((seg.i$end.pos - seg.i$start.pos)/1e6 , 0)
filter.size   <- (seg.i$end.pos - seg.i$start.pos) >= 10e6
avg.unlogR <- mean(sample.i$adjusted.ratio, na.rm = TRUE)
@

<<label=CPforSNP, eval = FALSE, results=hide>>=
CPsnp.example <- baf.model.fit(Bf = seg.i$Bf[filter.size],
                        depth.ratio = seg.i$depth.ratio[filter.size], 
                        weight.ratio = weights.snp[filter.size],
                        weight.Bf = weights.snp[filter.size],
                        avg.depth.ratio = avg.unlogR,
                        cellularity = seq(0.1,1,0.01), 
                        ploidy = seq(1,7,0.1),
                        priors.table = data.frame(CN = 2, value = 2))
@

<<label=loadCPforSNP, eval = TRUE, echo=FALSE>>=
data(CPsnp.example)
@
<<label=getCPparamSNP, eval = TRUE>>=
cint <- get.ci(CPsnp.example)

cellularity <- cint$max.y
ploidy   <- cint$max.x
@


\subsection{Cellularity and ploidy plot for SNP array}
\begin{figure}[H]
      \centering
<<label=CPplotSNP,fig=TRUE,width=6,height=6>>=
cp.plot(CPsnp.example)
cp.plot.contours(CPsnp.example, add = TRUE)
@ 
      \caption{Result from the Bayesian inference over the defined range of cellularity and ploidy from artificial SNP array data. The color indicate the log-likelihood of the corresponding cellularity/ploidy combinations.}
      \label{fig:cpPlotSNP}
\end{figure}

\subsection{Call for copy number variation using inferred parameters.}

<<label=bafBayesSNP, eval=TRUE>>=
snp.seg.cn <- baf.bayes(Bf = seg.i$Bf,
                        depth.ratio = seg.i$depth.ratio, 
                        avg.depth.ratio = avg.unlogR,
                        cellularity = cellularity,
                        weight.ratio = 2 * 300,
                        weight.Bf = 300, ratio.priority = FALSE,
                        ploidy = ploidy, CNt.max = 10)

segmented.snp <- cbind(seg.i, snp.seg.cn)

head(segmented.snp[segmented.snp$chromosome == 1, ])
@

\subsection{Graphical representation of copy number with SNP arrays}

\begin{figure}[H]
      \centering
<<label=chrViewNoMutCP1, fig=TRUE, width=6, height=6>>=

chromosome.view(baf.windows = snp.b.win[[1]], 
                ratio.windows = snp.r.win[[1]],  min.N.ratio = 1,
                segments = segmented.snp[segmented.snp$chromosome == "1", ],
                cellularity = cellularity, ploidy = ploidy,
                avg.depth.ratio = avg.unlogR, main = "1")
@
      \caption{Plots B-allele frequencies (top) and un-logged-logR (bottom) with SNP array data. Chromosome 16. Horizontal dotted line indicate different copy number/ allele state. }
      \label{fig:crhomViewNoMutCP1}
\end{figure}

\begin{figure}[H]
      \centering
<<label=chrViewNoMutCP16, fig=TRUE, width=6, height=6>>=
chromosome.view(baf.windows = snp.b.win[[16]], 
                ratio.windows = snp.r.win[[16]],  min.N.ratio = 1,
                segments = segmented.snp[segmented.snp$chromosome == "16", ],
                cellularity = cellularity, ploidy = ploidy,
                avg.depth.ratio = avg.unlogR, main = "16")
@
      \caption{Plots B-allele frequencies (top) and un-logged-logR (bottom) with SNP array data. Chromosome 16. Horizontal dotted line indicate different copy number/ allele state. }
      \label{fig:crhomViewNoMutCP16}
\end{figure}

\begin{figure}[H]
      \centering
<<label=genomeViewCNtSNP, fig=TRUE, width=7, height=3>>=

genome.view(seg.cn = segmented.snp, info.type = "CNt")
legend("bottomright", bty="n", c("Tumor copy number"),col = c("red"), 
       inset = c(0, -0.4), pch=15, xpd = TRUE)
@
      \caption{Genome whide copy number profile obtained from one SNP array.}
      \label{fig:genomeViewCNtSNP}
\end{figure}

\begin{figure}[H]
      \centering
<<label=genomeViewABSNP, fig=TRUE, width=7, height=3>>=
genome.view(seg.cn = segmented.snp, info.type = "AB")
legend("bottomright", bty = "n", c("A-allele","B-allele"), col= c("red", "blue"), 
       inset = c(0, -0.45), pch = 15, xpd = TRUE)
@
      \caption{Genome whide A anf B alleles profile, obtained from one SNP array.}
      \label{fig:genomeViewABSNP}
\end{figure}
%\newpage

\end{document}

                